[universal-docker] Docker buildx component missing or broken

Universal docker mod comes with a new version of Docker that should support the new Buildkit build system for creating docker images, however attempting to do so results in the following error:

```
$ DOCKER_BUILDKIT=1 docker build .                                             
ERROR: BuildKit is enabled but the buildx component is missing or broken.
       Install the buildx component to build images with BuildKit:
       https://docs.docker.com/go/buildx/
```
Using the Universal Docker plugin on codeserver Docker container.
Please let me know if I missed something or if there is any further information I can provide.

Thank you for the quick turnaround and apologies for the delay on my response.

The PR for the module fixes the lack of buildkit and the image starts building, but I'm running into an issue that I'm having trouble determining if the cause lies in codeserver/dockermod or just something about my setup/codebase.

It doesn't seem like it should be an issue with codeserver/dockermod, but the same Dockerfile runs fine on the exact same codebase on my desktop, which is why I'm on the fence. (it seems like the builder container is looking for something in the codeserver's filesystem??)

Here's the error, and if you agree that it doesn't seem like it should be codeserver/dockermod related, I'm cool with this issue being closed/merged, but someone else might want to double-check another repo/build to make fully sure that it's working as intended.

Thanks again!

```
...
 => [builder 4/5] COPY . .                                                                                                                                       26.7s 
 => ERROR [builder 5/5] RUN yarn build                                                                                                                            6.4s 
------                                                                                                                                                                 
 > [builder 5/5] RUN yarn build:                                                                                                                                       
4.438 node:internal/modules/cjs/loader:1051                                                                                                                            
4.438   throw err;                                                                                                                                                     
4.438   ^                                                                                                                                                              
4.438                                                                                                                                                                  
4.438 Error: Cannot find module '/config/workspace/project/node_modules/.store/next-virtual-2f7d86e049/node_modules/next/dist/bin/next'
4.438     at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
4.438     at Module._load (node:internal/modules/cjs/loader:901:27)
4.438     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
4.438     at node:internal/main/run_main_module:23:47 {
4.438   code: 'MODULE_NOT_FOUND',
4.438   requireStack: []
4.438 }
4.438 
4.438 Node.js v20.5.0
------
ERROR: failed to solve: executor failed running [/bin/sh -c yarn build]: exit code: 1 ```

--- a/Dockerfile
+++ b/Dockerfile
@@ -67,10 +67,10 @@ RUN \
   fi && \
   echo "**** grab buildx plugin ****" && \
   curl -fo \
-    /root-layer/docker-bins/buildx_x86_64 -L \
+    /root-layer/docker-bins/docker-buildx_x86_64 -L \
     "https://github.com/docker/buildx/releases/download/${BUILDX_RELEASE}/buildx-${BUILDX_RELEASE}.linux-amd64" && \
   curl -fo \
-    /root-layer/docker-bins/buildx_aarch64 -L \
+    /root-layer/docker-bins/docker-buildx_aarch64 -L \
     "https://github.com/docker/buildx/releases/download/${BUILDX_RELEASE}/buildx-${BUILDX_RELEASE}.linux-arm64" && \
   chmod +x /root-layer/docker-bins/* && \
   rm -rf /tmp/*

--- a/Dockerfile
+++ b/Dockerfile
@@ -5,10 +5,6 @@ FROM ghcr.io/linuxserver/baseimage-alpine:3.17 as buildstage
 ARG MOD_VERSION
 
 RUN \
-  echo "**** install packages ****" && \
-  apk add --no-cache \
-    git \
-    go && \
   echo "**** retrieve latest version ****" && \
   if [[ -z "${MOD_VERSION+x}" ]]; then \
     DOCKER_RELEASE=$(curl -sX GET "https://api.github.com/repos/moby/moby/releases/latest" \

--- a/Dockerfile
+++ b/Dockerfile
@@ -25,7 +25,6 @@ RUN \
   mkdir -p \
     /root-layer/docker-bins \
     /tmp/docker_x86_64 \
-    /tmp/docker_armv7l \
     /tmp/docker_aarch64 && \
   curl -fo \
     /tmp/docker_x86_64.tgz -L \
@@ -34,13 +33,6 @@ RUN \
     /tmp/docker_x86_64.tgz -C \
     /tmp/docker_x86_64 --strip-components=1 && \
   cp /tmp/docker_x86_64/docker /root-layer/docker-bins/docker_x86_64 && \
-  curl -fo \
-    /tmp/docker_armv7l.tgz -L \
-    "https://download.docker.com/linux/static/stable/armhf/docker-${DOCKER_RELEASE}.tgz" && \
-  tar xf \
-    /tmp/docker_armv7l.tgz -C \
-    /tmp/docker_armv7l --strip-components=1 && \
-  cp /tmp/docker_armv7l/docker /root-layer/docker-bins/docker_armv7l && \
   curl -fo \
     /tmp/docker_aarch64.tgz -L \
     "https://download.docker.com/linux/static/stable/aarch64/docker-${DOCKER_RELEASE}.tgz" && \
@@ -52,9 +44,6 @@ RUN \
   curl -fo \
     /root-layer/docker-bins/docker-compose_x86_64 -L \
     "https://github.com/docker/compose/releases/download/v${COMPOSE_RELEASE}/docker-compose-linux-x86_64" && \
-  curl -fo \
-    /root-layer/docker-bins/docker-compose_armv7l -L \
-    "https://github.com/docker/compose/releases/download/v${COMPOSE_RELEASE}/docker-compose-linux-armv7" && \
   curl -fo \
     /root-layer/docker-bins/docker-compose_aarch64 -L \
     "https://github.com/docker/compose/releases/download/v${COMPOSE_RELEASE}/docker-compose-linux-aarch64" && \
@@ -71,16 +60,18 @@ RUN \
   curl -fo \
     /root-layer/docker-bins/compose-switch_aarch64 -L \
     "https://github.com/docker/compose-switch/releases/download/v${SWITCH_RELEASE}/docker-compose-linux-arm64" && \
-  echo "**** compile compose switch on armhf ****" && \
+  echo "**** retrieve latest buildx version ****" && \
+  if [ -z ${BUILDX_RELEASE+x} ]; then \
+    BUILDX_RELEASE=$(curl -sX GET "https://api.github.com/repos/docker/buildx/releases/latest" \
+      | awk '/tag_name/{print $4;exit}' FS='[""]'); \
+  fi && \
+  echo "**** grab buildx plugin ****" && \
   curl -fo \
-    /tmp/compose-switch.tar.gz -L \
-    "https://github.com/docker/compose-switch/archive/refs/tags/v${SWITCH_RELEASE}.tar.gz" && \
-  mkdir -p /tmp/compose-switch && \
-  tar xf \
-    /tmp/compose-switch.tar.gz -C \
-    /tmp/compose-switch --strip-components=1 && \
-  cd /tmp/compose-switch && \
-  CGO_ENABLED=0 GOOS=linux GOARCH=arm go build -ldflags="-s -w -X github.com/docker/compose-switch/internal.Version=${SWITCH_RELEASE}" -o /root-layer/docker-bins/compose-switch_armv7l ./main.go && \
+    /root-layer/docker-bins/buildx_x86_64 -L \
+    "https://github.com/docker/buildx/releases/download/${BUILDX_RELEASE}/buildx-${BUILDX_RELEASE}.linux-amd64" && \
+  curl -fo \
+    /root-layer/docker-bins/buildx_aarch64 -L \
+    "https://github.com/docker/buildx/releases/download/${BUILDX_RELEASE}/buildx-${BUILDX_RELEASE}.linux-arm64" && \
   chmod +x /root-layer/docker-bins/* && \
   rm -rf /tmp/*
 

--- a/README.md
+++ b/README.md
@@ -1,6 +1,6 @@
 # Docker - Docker mod for all images
 
-This mod adds `docker` and `docker-compose` binaries to any linuxserver image.
+This mod adds the `docker` binary as wells as the `buildx` and the `compose` plug-ins to any linuxserver image.
 
 **IMPORTANT NOTE**: For docker access inside a container, a volume mapping needs to be added for `/var/run/docker.sock:/var/run/docker.sock:ro` in the container's docker run/create/compose. If you'd like to connect to a remote docker service instead, you don't have to map the docker sock; you can either set an env var for `DOCKER_HOST=remoteaddress` or use the docker cli option `-H`.
 

--- a/root/etc/s6-overlay/s6-rc.d/init-mod-universal-docker-setup/run
+++ b/root/etc/s6-overlay/s6-rc.d/init-mod-universal-docker-setup/run
@@ -2,10 +2,15 @@
 
 echo "**** installing docker and docker compose ****"
 ARCH=$(uname -m)
-if [ -d "/docker-bins" ] ; then
+if [[ "${ARCH}" = "armv7l" ]]; then
+  echo "**** The universal docker mod no longer supports arm32v7/armhf per https://info.linuxserver.io/issues/2023-07-01-armhf/ ****"
+  exit 0
+fi
+if [[ -d "/docker-bins" ]] ; then
   echo "Copying over docker and docker-compose binaries"
   mkdir -p /usr/local/lib/docker/cli-plugins
   mv "/docker-bins/docker-compose_${ARCH}" /usr/local/lib/docker/cli-plugins/docker-compose
+  mv "/docker-bins/buildx_${ARCH}" /usr/local/lib/docker/cli-plugins/docker-buildx
   mv "/docker-bins/docker_${ARCH}" /usr/local/bin/docker
   mv "/docker-bins/compose-switch_${ARCH}" /usr/local/bin/docker-compose
   rm -rf /docker-bins
@@ -13,19 +18,19 @@ else
   echo "**** docker and docker-compose already installed, skipping ****"
 fi
 
-if [ -S /var/run/docker.sock ]; then
+if [[ -S /var/run/docker.sock ]]; then
   DOCKER_GID=$(stat -c '%g' "/var/run/docker.sock")
   if id -G abc | grep -qw "$DOCKER_GID"; then
     exit 0
   else
     DOCKER_NAME=$(getent group "${DOCKER_GID}" | awk -F: '{print $1}')
-    if [ -z "${DOCKER_NAME}" ]; then
+    if [[ -z "${DOCKER_NAME}" ]]; then
       DOCKER_NAME="dockergroup"
       groupadd -g "${DOCKER_GID}" "${DOCKER_NAME}"
     fi
     usermod -aG "${DOCKER_NAME}" abc
   fi
-elif [ -n "$DOCKER_HOST" ]; then
+elif [[ -n "$DOCKER_HOST" ]]; then
   echo "**** Remote docker service $DOCKER_HOST will be used ****"
 else
   echo "**** Please map /var/run/docker.sock for access to docker service on host. Alternatively you can manually define a remote host address with the docker cli option -H ****"

--- a/root/etc/s6-overlay/s6-rc.d/init-mod-universal-docker-setup/run
+++ b/root/etc/s6-overlay/s6-rc.d/init-mod-universal-docker-setup/run
@@ -10,7 +10,7 @@ if [[ -d "/docker-bins" ]] ; then
   echo "Copying over docker and docker-compose binaries"
   mkdir -p /usr/local/lib/docker/cli-plugins
   mv "/docker-bins/docker-compose_${ARCH}" /usr/local/lib/docker/cli-plugins/docker-compose
-  mv "/docker-bins/buildx_${ARCH}" /usr/local/lib/docker/cli-plugins/docker-buildx
+  mv "/docker-bins/docker-buildx_${ARCH}" /usr/local/lib/docker/cli-plugins/docker-buildx
   mv "/docker-bins/docker_${ARCH}" /usr/local/bin/docker
   mv "/docker-bins/compose-switch_${ARCH}" /usr/local/bin/docker-compose
   rm -rf /docker-bins

